### PR TITLE
[#872] feat(tez): Add the common and utils class 

### PR DESCRIPTION
**Pull request by lifeSo**
### What changes were proposed in this pull request?

Tez Shuffle Read supporting Remote Shuffle related common and utils

### Why are the changes needed?

Fix:  https://github.com/apache/incubator-uniffle/issues/872

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

UT.

### CODE REVIEW

File: client-tez/src/test/java/org/apache/tez/common/RssTezUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class RssTezUtilsTest {
+
+  private static final String EXPECTED_EXCEPTION_MESSAGE = "Exception should be thrown";
+
+//  @Test
```
Comment (jerqi) [inline client-tez/src/test/java/org/apache/tez/common/RssTezUtilsTest.java:41]: Could we remove this unused comment code?

**Comment by jerqi** [comment]
This pr is also a little large. Could you split into more smaller prs?

File: client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+
+  public static int computeShuffleId(int tezDagID, int upTezVertexID, int downTezVertexID) {
+    return tezDagID * (SHUFFLE_ID_MAGIC * SHUFFLE_ID_MAGIC)  + upTezVertexID * SHUFFLE_ID_MAGIC + downTezVertexID;
+  }
+
+  public static int computeShuffleId(int tezDagID, int tezVertexID) {
+    return tezDagID * 10000 + tezVertexID;
+  }
+
+  private static final int VERTEX_ID_MAPPING_MAX_ID = 500;
```
Comment (jerqi) [inline client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java:291]: Could we put the constants in the top place of this class?

File: client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+    } else if (VERTEX_ID_MAPPING_REDUCER.equals(ss[0])) {
+      return 600 + Integer.parseInt(ss[1]);
+    } else {
+      throw new RssException("Wrong vertex name to id mapping, vertexName:" + vertexName);
+    }
+  }
+
+
+
+  private static int SHUFFLE_ID_MAGIC = 1000;
```
Comment (jerqi) [inline client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java:311]: Could we use `final` key word herer?

File: client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+          RssTezConfig.RSS_CLIENT_RETRY_MAX,
+          retryMax,
+          RssTezConfig.RSS_CLIENT_RETRY_INTERVAL_MAX,
+          retryIntervalMax,
+          RssTezConfig.RSS_CLIENT_SEND_CHECK_TIMEOUT_MS,
+          sendCheckTimeout));
+    }
+  }
+
+  public static int computeShuffleId(int tezDagID, String upVertexName, String downVertexName) {
```
Comment (jerqi) [inline client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java:274]: Could you add some comments to explain how to generate the shuffleID according to vertex information?

File: client-tez/src/main/java/org/apache/tez/common/TezClassLoader.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+ *
+ * It uses URLClassLoader with this class' classloader as parent classloader.
+ * And hence first delegates the resource loading to parent and then to the URLs
+ * added. The process must be setup to use by invoking setupTezClassLoader() which sets
+ * the global TezClassLoader as current thread context class loader. All threads
+ * created will inherit the classloader and hence will resolve the class/resource
+ * from TezClassLoader.
+ */
+
+public class TezClassLoader extends URLClassLoader {
```
Comment (jerqi) [inline client-tez/src/main/java/org/apache/tez/common/TezClassLoader.java:39]: Why do need this class? Could we reuse Tez's implement?

File: pom.xml [hunk 6]
```diff
@@ -1625,10 +1625,13 @@
       </dependencyManagement>
     </profile>
     <profile>
-      <id>tez</id>
+      <id>tez0.9</id>
```
Comment (jerqi) [inline pom.xml:1628]: If we only have one version Tez, we don't need the name tez + version.

File: client-tez/src/main/java/org/apache/tez/common/TezClassLoader.java (hunk 5 above)
Comment (lifeSo) [reply client-tez/src/main/java/org/apache/tez/common/TezClassLoader.java:39]: Install the tez class loader, which can be used add new resources

click below to refer more:

https://github.com/apache/tez/blob/ccad71dabceb850b58ff38e77fd206e68f62ce4b/tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java#L2338

File: client-tez/pom.xml [hunk 7]
```diff
@@ -1,86 +1,91 @@
...
-  -->
 <project xmlns="http://maven.apache.org/POM/4.0.0"
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <modelVersion>4.0.0</modelVersion>
     <parent>
         <artifactId>uniffle-parent</artifactId>
         <groupId>org.apache.uniffle</groupId>
-        <version>0.8.0-SNAPSHOT</version>
+        <version>0.7.0</version>
```
[the remark this answers is not in the document]
Comment (lifeSo) [reply client-tez/pom.xml]: Ok, we corrected it.

File: pom.xml (hunk 6 above)
Comment (lifeSo) [reply pom.xml]: Ok, we corrected it.

File: client-tez/src/test/java/org/apache/tez/common/RssTezUtilsTest.java (hunk 1 above)
Comment (lifeSo) [reply client-tez/src/test/java/org/apache/tez/common/RssTezUtilsTest.java]: ok

File: client-tez/src/test/java/org/apache/tez/common/TezIdHelperTest.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package org.apache.tez.common;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TezIdHelperTest {
+
+  @Test
+  public void TestTetTaskAttemptId(){
+    TezIdHelper tezIdHelper = new TezIdHelper();
```
[the remark this answers is not in the document]
Comment (lifeSo) [reply client-tez/src/test/java/org/apache/tez/common/TezIdHelperTest.java:10]: ok and done

File: client-tez/src/main/java/org/apache/tez/common/GetShuffleServerResponse.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+package org.apache.tez.common;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.util.Objects;
+
+import org.apache.hadoop.io.Writable;
+
+public class GetShuffleServerResponse implements Writable {
```
[the remark this answers is not in the document]
Comment (lifeSo) [reply client-tez/src/main/java/org/apache/tez/common/GetShuffleServerResponse.java]: ok, we removed the file in this commit

File: client-tez/src/test/java/org/apache/tez/common/TezIdHelperTest.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package org.apache.tez.common;
```
Comment (jerqi) [inline client-tez/src/test/java/org/apache/tez/common/TezIdHelperTest.java:1]: Could you add the Apache license?

File: client-tez/pom.xml [hunk 11]
```diff
@@ -176,4 +181,4 @@
             </plugin>
         </plugins>
     </build>
-</project>
+</project>
```
Comment (jerqi) [inline client-tez/pom.xml:184]: We should add a new line, otherwise there will be a warning.

File: client-tez/pom.xml [hunk 12]
```diff
@@ -1,20 +1,4 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!--
```
Comment (jerqi) [inline client-tez/pom.xml:2]: Why do you remove the license? we must include the Apache License in every file.

File: client-tez/src/main/java/org/apache/tez/common/TezClassLoader.java (hunk 5 above)
Comment (lifeSo) [reply client-tez/src/main/java/org/apache/tez/common/TezClassLoader.java]: We removed the file in this commit, battle it in next MR/PR

File: client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java (hunk 4 above)
Comment (lifeSo) [reply client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java:302]: Ok, we added it and commit it.

File: client-tez/pom.xml (hunk 12 above)
Comment (lifeSo) [reply client-tez/pom.xml]: ok, we added it

File: client-tez/pom.xml (hunk 11 above)
Comment (lifeSo) [reply client-tez/pom.xml]: ok, we add added it

File: client-tez/src/test/java/org/apache/tez/common/TezIdHelperTest.java (hunk 10 above)
Comment (lifeSo) [reply client-tez/src/test/java/org/apache/tez/common/TezIdHelperTest.java:18]: ok, we added it

File: client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java [hunk 13]
```diff
@@ -0,0 +1,470 @@
...
+      throw new RssException("appAttemptId  " + appAttemptId + " is wrong");
+    }
+    long highBytes = (long)taskAttemptID - (appAttemptId - 1) * 1000;
+    if (highBytes > MAX_ATTEMPT_ID || highBytes < 0) {
+      throw new RssException("TaskAttempt " + taskAttemptID + " high bytes " + highBytes + " exceed");
+    }
+    return (highBytes << (Constants.TASK_ATTEMPT_ID_MAX_LENGTH + Constants.PARTITION_ID_MAX_LENGTH)) + lowBytes;
+  }
+
+  public static Roaring64NavigableMap fetchAllRssTaskIds(Set<InputAttemptIdentifier> successMapTaskAttempts,
```
[the remark this answers is not in the document]
Comment (lifeSo) [reply client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java]: ok, we corrected it

File: client-tez/pom.xml [hunk 14]
```diff
@@ -26,61 +27,81 @@
...
     <artifactId>rss-client-tez</artifactId>
     <version>0.8.0-SNAPSHOT</version>
     <packaging>jar</packaging>
     <name>Apache Uniffle Client (Tez)</name>
 
     <dependencies>
         <dependency>
-            <groupId>org.apache.uniffle</groupId>
-            <artifactId>rss-client</artifactId>
+            <groupId>org.apache.tez</groupId>
```
Comment (jerqi) [inline client-tez/pom.xml:37]: Could we remove the `version`? The pom will extend the parent tez version. We use the parent pom to manage the third party package version.

File: client-tez/pom.xml [hunk 15]
```diff
@@ -26,61 +27,81 @@
         <relativePath>../pom.xml</relativePath>
     </parent>
 
-    <groupId>org.apache.uniffle</groupId>
```
Comment (jerqi) [inline client-tez/pom.xml:29]: Why do we remove this groupId? Could we keep consistent style with other submodule pom files?

**Comment by jerqi** [comment]
cc @zhengchenyu Do you have another suggestion?

File: client-tez/pom.xml (hunk 15 above)
Comment (lifeSo) [reply client-tez/pom.xml]: ok, we reverted it.

File: client-tez/pom.xml [hunk 16]
```diff
@@ -34,53 +35,74 @@
...
+            <artifactId>tez-common</artifactId>
+            <version>${tez.version}</version>
         </dependency>
-
         <dependency>
-            <groupId>org.apache.commons</groupId>
-            <artifactId>commons-lang3</artifactId>
+            <groupId>org.apache.tez</groupId>
+            <artifactId>tez-runtime-library</artifactId>
+            <version>${tez.version}</version>
```
Comment (jerqi) [inline client-tez/pom.xml:45]: ditto.

File: client-tez/pom.xml [hunk 17]
```diff
@@ -34,53 +35,74 @@
...
         </dependency>
         <dependency>
-            <groupId>net.jpountz.lz4</groupId>
-            <artifactId>lz4</artifactId>
+            <groupId>org.apache.tez</groupId>
+            <artifactId>tez-runtime-internals</artifactId>
+            <version>${tez.version}</version>
```
Comment (jerqi) [inline client-tez/pom.xml:50]: ditto.

File: client-tez/pom.xml [hunk 18]
```diff
@@ -34,53 +35,74 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.tez</groupId>
+            <artifactId>tez-dag</artifactId>
+            <version>${tez.version}</version>
```
Comment (jerqi) [inline client-tez/pom.xml:55]: ditto.

File: client-tez/pom.xml [hunk 19]
```diff
@@ -34,53 +35,74 @@
...
+            </exclusions>
         </dependency>
-
         <dependency>
-            <groupId>com.google.guava</groupId>
-            <artifactId>guava</artifactId>
-            <version>${guava.version}</version>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-common</artifactId>
+            <version>${hadoop.version}</version>
```
Comment (jerqi) [inline client-tez/pom.xml:71]: ditto.

File: client-tez/pom.xml [hunk 20]
```diff
@@ -34,53 +35,74 @@
...
-            <groupId>org.mockito</groupId>
-            <artifactId>mockito-core</artifactId>
-            <scope>test</scope>
-        </dependency>
-        <dependency>
-            <groupId>com.github.luben</groupId>
-            <artifactId>zstd-jni</artifactId>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-mapreduce-client-core</artifactId>
+            <version>${hadoop.version}</version>
```
Comment (jerqi) [inline client-tez/pom.xml:95]: ditto.

File: client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java [hunk 21]
```diff
@@ -0,0 +1,451 @@
...
+
+  public static String uniqueIdentifierToAttemptId(String uniqueIdentifier) {
+    if (uniqueIdentifier == null) {
+      throw new RuntimeException("uniqueIdentifier should not be null");
+    }
+    String[] ids = uniqueIdentifier.split("_");
+    return StringUtils.join(ids, "_", 0, 7);
+  }
+
+  public static int getInt(Configuration rssJobConf, Configuration mrJobCOnf, String key, int defaultValue) {
```
Comment (zhengchenyu) [inline client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java:175]: 1. name fix
rssJobConf -> rssConf, mrJobCOnf -> tezConf
2. In client-mr mode, we use rss_conf.xml to transmit rss config. In client-tez mode, I think rss_conf.xml is unnecessary, so we can retain only one config in parameters.

File: client-tez/pom.xml (hunk 14 above)
Comment (lifeSo) [reply client-tez/pom.xml]: ok, we removed it and pushed it.

File: client-tez/pom.xml (hunk 16 above)
Comment (lifeSo) [reply client-tez/pom.xml]: ok, we removed version config and pushed the code.

File: client-tez/pom.xml (hunk 17 above)
Comment (lifeSo) [reply client-tez/pom.xml]: ok, we removed version config and pushed the code.

File: client-tez/pom.xml (hunk 18 above)
Comment (lifeSo) [reply client-tez/pom.xml]: ok, we removed version config and pushed the code.

File: client-tez/pom.xml (hunk 19 above)
Comment (lifeSo) [reply client-tez/pom.xml]: ok, we removed version config and pushed the code.

File: client-tez/pom.xml (hunk 20 above)
Comment (lifeSo) [reply client-tez/pom.xml]: ok, we removed version config and pushed the code.

File: client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java (hunk 21 above)
Comment (lifeSo) [reply client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java]: ok, we removeed the getInt method()

Comment (lifeSo) [reply client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java]: ok, we removed the getInt method()

**Review by jerqi — APPROVED** [review]
LGTM, thanks @lifeSo @zhengchenyu. Merged to master, wait for CI.

**Pull request merged**